# PPL Toolbox
A pipeline to process the multi-interaction datasets based on 3C and long reads sequencing.
![image](https://github.com/versarchey/PPL/blob/main/logo.png)

---


## Install
dependencies：

running environment: ubuntu(>=18.04)/centos(>=6), jdk(>=1.8), minimap2

## Extract contacts
###For experiments with restriction enzyme digestion step
####step1 virtual restriction
The simulated digest file will be generated in this step and can be reused, adding this file with the --restrictionsiteFile parameter avoids duplicate calculations.
模拟酶切文件将在这一步生成，该文件可以复用，将该文件使用--restrictionsiteFile参数添加可以避免重复计算.
####step2 mapping to genome, assigning
virtual fragment and generating aligntable;
This step starts with matching the reads to the reference genome using a comparator (minimap2), which is currently the best comparator for pore-c reads in the current test results.
You can also use --skipmap Y to skip the comparison process, which requires you to place the bam in the output directory beforehand and name it in the canonical PREFIX.bam fashion.
This process will recognize the file and use it to generate an aligntable file and use it for subsequent analysis (see the example below). Here the aligntable will be annotated using the digest file generated by step1.
这一步首先会用比对器(minimap2)将reads比对到参考基因组，minimap2在当前的测试结果下是目前比对pore-c reads效果最优的比对器。
你也可以使用--skipmap Y来跳过比对过程，这需要你将bam事先放置到output目录并以规范的PREFIX.bam方式命名。
这个流程会识别该文件并将其用于生成aligntable文件并用于后续分析（请查看下面的示例）。这里aligntable会使用step1产生的酶切文件注释。
####step3 classification, all mapping records are tagged with different types
|Enzyme Name|            Motif|
| :---: | :---: |
|HindIII      |           A^AGCTT|
|MboI|^GATC|
|BglII|A^GATCT|
|DpnII       |            ^GATC|
|Sau3AI|^GATC|
|NlaIII         |         CATG^|
|Hinf1|G^ANTC|
|AluI|AG^CT|
|NcoI          |          C^CATGG|
####step4 extracting gaps from raw reads;
####step5 mapping gaps to genome;
####step6 filtering gaps.alignmentable;
####step7 merging results.




## Examples 示例
###0 Help 使用说明
    java -jar PPL.jar --help -Y

###1 Virtual Restriction 生成模拟酶切文件
    #Usage
    java -cp PPL.jar utils.VirDigestTool <genomeFile> <enzyme/resSite> <result.bed>

    #for example
    java -cp PPL.jar utils.VirDigestTool \
        ./hg38.fa \
        ./CATG^ \
        ./hg38.nlaiii.res.bed

    #results like below
    ##chr, start, end, length, index
    chrM    0       47      47      0
    chrM    47      697     650     1
    chrM    697     1345    648     2
    chrM    1345    1569    224     3
    chrM    1569    2317    748     4
    chrM    2317    2440    123     5
    chrM    2440    2560    120     6
    chrM    2560    2855    295     7
    chrM    2855    3315    460     8
    chrM    3315    3842    527     9
####Some Enzymes and their motifs 常用的内切酶
|Enzyme Name|            Motif|
| :---: | :---: |
|HindIII      |           A^AGCTT|
|MboI|^GATC|
|BglII|A^GATCT|
|DpnII       |            ^GATC|
|Sau3AI|^GATC|
|NlaIII         |         CATG^|
|Hinf1|G^ANTC|
|AluI|AG^CT|
|NcoI          |          C^CATGG|



### 2 Processing (Mapping Reads to Reference and Filtering alignments)  处理

[comment]: <> (|Experiment type|	Restriction site file|	no_balance|)

[comment]: <> (| :---: | :---: |:---:|)

[comment]: <> (|in situ Hi-C	|yes	|False|)

[comment]: <> (|dilution Hi-C|	yes	|False|)

[comment]: <> (|DNase Hi-C|no	|False|)

[comment]: <> (|Micro-C|	no	|False|)

[comment]: <> (|Capture Hi-C|	yes	|True|)

[comment]: <> (|ChIA-PET|	no	|True|)

####2.1 Directly Mapping(DM) mode (Recommended) 直接比对
DM mode is our recommended mode to process data from restriction-enzyme-based 3C experiment.

    #some directories and files
    work_dir="/public/home/zjwang/analysis/pore-c/test"
    fq_dir="/public/home/zjwang/data/pore-c"
    genome="/public/home/zjwang/analysis/pore-c/hg38_bwaindex/hg38.fa"
    res_d="/public/home/zjwang/analysis/pore-c/hg38_bwaindex/hg38.nlaiii.res.bed"
    id="SRR11589402"

    #processing from raw .fastq file
    ##a simple example
    java -jar ../PPL.jar \ 
        --ligation_type res \
        --genomefile $genome \
        --fastq ${fq_dir}/${id}/${id}_1.fastq.gz \
        --output ./ \
        --prefix $id \
        --start_step 2 \
        --restrictionsiteFile ${res_d} \
        --thread 12
    ##a complex example
    java -jar ../PPL.jar --ligation_type res \
        --genomefile $genome \
        --fastq ${fq_dir}/${id}/${id}_1.fastq.gz \
        --splitReads N --resRemove N --disRemove N \
        --output ./ \
        --prefix $id \
        --skipmap N \ (do not skip mapping step)
        --start_step 2 \
        --stop_step 4 \
        --restrictionsiteFile ${res_d} \
        --thread 12 \
        --cutoffMapq 1 \
        --weightAS 1 \
        --filter res
    
    #processing from .bam file
    ##a simple example
    java -jar ../PPL.jar \ 
        --ligation_type res \
        --genomefile $genome \
        --fastq ${fq_dir}/${id}/${id}_1.fastq.gz \
        --output ./ \
        --prefix $id \
        --start_step 2 \
        --restrictionsiteFile ${res_d} \
        --thread 12
    ##a complex example
    java -jar ../PPL.jar --ligation_type res \
        --genomefile $genome \
        --fastq ${fq_dir}/${id}/${id}_1.fastq.gz \
        --splitReads N --resRemove N --disRemove N \
        --output ./ \
        --prefix $id \
        --skipmap Y \ (skip mapping step)
        --start_step 2 \
        --stop_step 4 \
        --restrictionsiteFile ${res_d} \
        --thread 12 \
        --cutoffMapq 1 \
        --filter res
####2.2 Splitting reads and Mapping(SM) mode 按酶切识别位点或linker序列断开后比对
Depending on the experiment type, our pipeline uses different methods to split reads from restriction-enzyme-based (like in-situ Hi-C) and linker-based experiments (like ChIA-PET) respectively.

    #To restriction-enzyme-based experiment (Unrecommended)
    ##splitting reads
    java -cp PPL.jar utils.SplitByRes \
        ../data/SRR11589402/SRR11589402_1.fastq.gz \
        ../data/SRR11589402/SRR11589402_1.splited.fastq.gz \
        nlaiii
    ##mapping split reads and filtering results
    java -jar ../PPL.jar \
        --fastq ../data/SRR11589402/SRR11589402_1.splited.fastq.gz \
        --ligation_type res \
        --genomefile ../hg38_bwaindex/hg38.chr.fa \ 
        --GENOME_INDEX ../hg38_bwaindex/hg38.chr.fa \
        --resRemove Y \
        --output ./ \
        --prefix SRR11589402_split \
        --skipmap N \
        --start_step 2 \
        --restrictionsiteFile ../hg38_bwaindex/hg38.chr.nlaiii.bed \
        --splitReads Y \
        --thread 12

    #To linker-based experiment (Testing)


## Arguments 参数含义

## Other useful tools 一些其他工具
    jar_path=${PATH_TO_JAR}/PPL.jar
### Data quality evaluation 数据质量评估
Detailed Illustration & Usage: [/generate-docx/README.md](https://github.com/versarchey/PPL/tree/main/generate-docx)

    #输出比对上的总长度占比
    java -cp ${jar_path} utils.CoverageReads <alignmenttable> <coverage>
    or
    java -cp ${jar_path} utils.CoverageReads <contacts> <coverage>
    
    #--output: (length of all valid frags / length of all mapped reads) = 0.8871755749913668
    #--outputfile content:
        #reads  length  lenValidMappings    coverageRatio   regionValidMappings
        SRR19920179.1   5836    5149    0.8822823851953393      [281..922),[1324..5832)
        SRR19920179.2   3856    3261    0.845695020746888       [29..1848),[2043..3485)
        SRR19920179.3   3753    3715    0.9898747668531841      [25..3740)
        SRR19920179.4   5952    4853    0.8153561827956989      [132..1981),[1993..2821),[3552..4874),[5032..5886)
        SRR19920179.5   5984    5909    0.9874665775401069      [56..3029),[3035..5971)
        SRR19920179.6   3004    2845    0.9470705725699068      [152..2997)    
        ......

    #计算比对边缘和酶切片段边缘的距离
    java -cp ${jar_path} utils.BoundaryCheck <monomers> <RestrictionSite> <distanceAllReads>
    java -cp ${jar_path} utils.BoundaryCheck2 <text(format:<chrName:start:end:status>)> <RestrictionSite> <distanceAllReads>
    
    #--outputfile content (.bc):
        #mapIndex   avgDist shorterDist 
        2       0       4
        1       5       5
        ......
    #--outputfile content (.bc.evenByChr):
        #chr    numFrags    numMapps    AvgDis  AvgDisMin
        chr9    297448  141252  17      3
        chr7    390397  181805  17      3
        ......

    #多交互的维数分布
    java -cp ${jar_path} utils.StatDimensionDistribution <contactFile> <outFile>
    #or
    java -cp ${jar_path} utils.StatDimensionDistribution <contactFile> <outFile>

    #按区间(bin)统计交互维数
    java -cp ${jar_path} utils.StatsDistributionByBin <monomers> <prefixOutputFiles> <chromSizes> <binSize, 1000000>

    #VPCs交互距离统计 inter-intra
    java -cp ${jar_path} utils.StatIntraInter <contacts> <stats.out>
    java -cp ${jar_path} utils.StatIntraInter2 <pairs> <stats.out>
    
    #--outputfile content :
        group   intra short-range(<=20kb)       intra long-range(>20kb) inter
        All     743371  5526092 2079079
        [2, 2]  17685   72429   20721
        [3, 3]  52842   261590  74746
        [4, 4]  91556   524933  154049
        [5, 6]  232155  1561497 497028
        [7, 11] 303240  2600146 1006767
        [12, 21]        45581   501516  319003
        [22, 2147483647]        312     3981    6765

    #VPCs的TAD检验
    java -cp ${jar_path} utils.StatByTAD <contacts> <domainfile(3 colmns)> <stats.out>

### Denoise multi-contacts 多交互降噪
Detailed Illustration & Usage: [/Denoise/README.md](https://github.com/versarchey/PPL-Toolbox/tree/main/Denoise/README.md)

    #Simple usage
    java -cp PPL.jar utils.FilterHyper
        #required
        <Input file: multi-contacts>
        <Output file: multi-contacts filtered>
        <Chrom size file> <binSize, e.g. 1000000>
        #optional
        <Percentile cutoff, default:0.85> 
        <Range, e.g. chr1: 1000000-2000000>


### Haplo-tagging contacts using Phased SNPs (only for diploid organism) 单倍型多交互区分（目前只支持二倍体）
Detailed Illustration & Usage: [/HaploTag/README.md](https://github.com/versarchey/PPL-Toolbox/tree/main/HaploTag/README.md)

    #1. 取前num_line行的结果进行预分型，绘制分类性能图谱供用户调参使用
    java -cp $jar_file utils.GenotypingQC
        -i <.bam> -v <PhasedSNP.vcf> -o <QCReport.txt>
        -l <num_line>
    
    #2. 正式分型
    java –cp $jar_file utils.GenotypingBySNPOld
        #required
        -i <.bam> -v <PhasedSNP.vcf> -o <PhasedRecords.txt>
        #optional
        --sampleName <TEXT>
        --baseq <num: 0-93>
        --mapq <num: 0-60> …
    
    #3. 根据pore-c的空缺值插补
    java –cp $jar_file utils.GenotypingImputation
        #required
        -i <.contacts> -v <PhasedRecords.vcf> -o <PhasedContacts.txt>

### Multi-way contacts Visualization 多交互可视化
Detailed Installation & Usage: [/visualization_Pore-C-track/README.md](https://github.com/versarchey/PPL-Toolbox/tree/main/visualization_Pore-C-track/README.md)

The visualization method is based on the pyGenomeTracks. You need to istall
it and copy the PoreTrack.py into pygenometracks/tracks folder. Then you can
add Pore-C track into your figure.

### Deduplicate amplicons (Only for single-cell experiments using PCR, such as scNanoHi-C)
    #去冗余 deduplicate PCR (only for scNanoHi-C)
    java -cp ${jar_path} utils.DeDuplicateByFrag <contactFile> <contactFile.dedup> <statsFile>
Detailed Illustration & Usage: [/PCR_deduplication/README.md](https://github.com/versarchey/PPL/blob/main/PCR_deduplication/README.md)


### Multi-way Contacts Annotation 多交互功能注释及分析
    Usage: 
        # Recommended
        java -cp $jar -cp PPL.jar utils.Contact2PCluster <.contacts file> <.pcluster file>
        java -cp $jar utils.AnnotatePClusters
            #required args
                -pc <pclusters>
                    #e.g. SRR19920179.54  chr6:44637890-44639773 … chr6:44498771-44499301
                -o <annotated pclusters>
                    # e.g. SRR19920179.54  chr6:44637890-44639773 … chr6:44498771-44499301 <anno info col>
                -a <annotation1,…>  
                    # e.g. chr9  85546539  85805188  -  AGTPBP1
                -at <col id,…>  
                    # e.g. 5
            #optional args
                --overlaps 	#输出overlap的长度

    Or:
        java -cp $jar utils.AnnotateContacts
            #required args
                -c <contacts file>
                -o <annotated contacts file>
                -a <annotation1,…>  
                -at <col id,…>  
            #optional args
                --overlaps

### Convert Format 格式转换
    #Decompose multi-way contacts to 4DN pairs (4DN)
    java -cp ${jar_path} utils.Contact2Pairs <monomers> <pairs>

    #Decompose multi-way contacts to Matrix (hicpro)
    java -cp ${jar_path} utils.Contact2Matrix <chromSize> <monomers> <outFile> <binSize>

    #Convert multi-way contacts to Hic medium (juicer)
    java -cp ${jar_path} utils.Contact2HicMedium <monomers> <hic_medium>

    #Convert multi-way contacts to cluster (SPRITE)
    java -cp ${jar_path} utils.Contact2Cluster <monomers> <cluster>

    #Convert multi-way contacts to Mbed (coloci analysis)
    java -cp ${jar_path} utils.Contact2Mbed <monomers> <mbed>

### Others
    #生成酶切位点文件 (step1)
    java -cp ${jar_path} utils.VirDigestTool <genomeFile> <enzyme/resSite> <result.bed>
 
    #按酶切位点切分fastq文件 (step0)
    java -cp ${jar_path} utils.SplitByRes <fastq> <fastq.splited> <ResSite>

    #.bed (需要cigar序列)转 aligntable文件
    java -cp ${jar_path} utils.Bed2Contact <.cigar_bed> <aligntable>

    #分配酶切片段
    java -cp ${jar_path} utils.AssignFragment <monomers> <RestrictionSite> <outputFile> <cutoff>(eg. 20 or 0.5, default 20) <a/b>(a:1 fragment at least; b:0 possibly. default 1;)
    sh fragmentAssign.sh <monomers> <enzymeSites.bed> <overlapCoverage eg.0.5>

    #提取没有比对到基因组上的区间
    java -cp ${jar_path} utils.ExtractGaps <original.fastq> <gaps.fastq> <coverageFile> <minLength>(default 50)

    #按阈值合并相邻的比对
    java -cp ${jar_path} utils.ContactMerge <contacts> <outputFile> <cutoff1> <cutoff2>


##结果文件格式格式
| Col 列号 | Field 字段 |Description 含义|示例|
| :---: | :---: | --- |---|
|1|chr|染色体id(String)|chr1|
|2|start|基因组上的起始位置(Number)|1000000|
|3|end|基因组上的终止位置(Number)|1003000|
|4|readID|reads id(String)|SRR11589402.5|
|5|readStart|reads上的起始位置(Number)|0|
|6|readEnd|reads上的终止位置(Number)|2900|
|7|readLength|reads长度(Number)|4000|
|8|score|mapq&AS(String)|60:2600|
|9|strand|strand(String)|+|
|10|mappingId|mappings id(Number)|100|
|11|status|mappings status(String)|passed|
|12|fragmentAssigned|匹配到的酶切片段|chr8-100890013-100890075,6903240:chr8-100890075-100890232|

    #chr, start, end, readID, readStart, readEnd, readLength, mapq:AS, strand, mapping id, status, (optional)fragment information
    chr12   92770602        92770829        SRR11589402.5   1237    1467    1517    15      -       0
    chr12   92396448        92396589        SRR11589402.5   780     910     1517    4       -       1
    chr17   35702412        35702739        SRR11589402.8   878     1196    1397    30      +       2
    chr12   80350217        80350338        SRR11589402.24  91      209     1086    28      +       3
    chr12   80935658        80935784        SRR11589402.24  1       127     1086    27      -       4
    chr12   80406876        80407054        SRR11589402.24  721     909     1086    12      +       5
    chr12   80406538        80406647        SRR11589402.24  251     358     1086    8       +       6
    chr10   25539254        25539714        SRR11589402.28  245     675     821     41      -       7
    chr8    100890038       100890232       SRR11589402.28  36      236     821     24      -       8
