# Denoise promiscuous muliti-contacts by filtering hypergraph

## Why
In our tests, we found that the percentage of VPCs in intra long-range and inter is significantly higher than in Hi-C and ChIA-PET in almost all the datasets of Pore-C and similar experiments published so far (Deshpande et al., 2022 ; Li et al., 2023 ; Zhong et al. al., 2023). In Hi-C data analysis, a high percentage of inter is often a reflection of poorer data quality (Lajoie et al., 2015).

<img src="https://github.com/versarchey/PPL-Toolbox/blob/main/Denoise/figs/compair_ratio.png" width = "80%" align=center />

Therefore, in order to trace the source of these long-range interactions to understand their formation, we made a sorted classification of GM12878 Hi-Pore-C interactions by order from low to high. The results show that the VPCs of inter type are mainly from ultra-high order interactions.

<img src="https://github.com/versarchey/PPL-Toolbox/blob/main/Denoise/figs/compair_ratio_by_dimension.png" width = "80%" align=center />

There may be multiple reasons for this phenomenon. First, random linkage events occur frequently in neighborhood linkage-based experiments, which have been quantitatively evaluated by previous authors in the ChIA-PET analysis process (Li et al., 2010). Compared to Hi-C, the linker (concatenator) of Pore-C has a more complex composition and needs to undergo multiple linkage steps throughout the formation process. Therefore, the higher the order of the concatemer the higher the probability of the participation artifacts. Second, the source of noise may also come from the ONT sequencing platform itself.2022 Yu et al. have systematically analyzed the preference of the ONT isochronous read sequencing platform in sequencing cfDNA, and found that a significant portion (6.3%) of ONT sequencing reads are in the form of chimeras (Yu et al., 2022). In addition, the method of VPCs generation amplifies the noise. Studies to date have tended to split VPCs using combinatorial methods that treat multiple interactions as complete graphs (Deshpande et al., 2022 ; Li et al., 2023), which leads to an exponential increase in the number of VPCs generated by higher-order interactions with increasing order, and thus ultra-high-order interactions are split into a large number of VPCs to amplify noise. 

<img src="https://github.com/versarchey/PPL-Toolbox/blob/main/Denoise/figs/generation.png" width = "80%" align=center />

### Reference
> Deshpande, A. S., Ulahannan, N., Pendleton, M., Dai, X., Ly, L., Behr, J. M., Schwenk, S., Liao, W., Augello, M. A., & Tyer, C. (2022). Identifying synergistic high-order 3D chromatin conformations from genome-scale nanopore concatemer sequencing. Nature Biotechnology, 1-12. 
> Li, W., Lu, J., Lu, P., Gao, Y., Bai, Y., Chen, K., Su, X., Li, M., Liu, J. e., Chen, Y., Wen, L., & Tang, F. (2023). scNanoHi-C: a single-cell long-read concatemer sequencing method to reveal high-order chromatin structures within individual cells. Nature methods. https://doi.org/10.1038/s41592-023-01978-w 
> Zhong, J.-Y., Niu, L., Lin, Z.-B., Bai, X., Chen, Y., Luo, F., Hou, C., & Xiao, C.-L. (2023). High-throughput Pore-C reveals the single-allele topology and cell type-specificity of 3D genome folding. Nature communications, 14(1), 1250. https://doi.org/10.1038/s41467-023-36899-x 
> Rao, Suhas S. P., Huntley, Miriam H., Durand, Neva C., Stamenova, Elena K., Bochkov, Ivan D., Robinson, James T., Sanborn, Adrian L., Machol, I., Omer, Arina D., Lander, Eric S., & Aiden, Erez L. (2014). A 3D Map of the Human Genome at Kilobase Resolution Reveals Principles of Chromatin Looping. Cell, 159(7), 1665-1680. https://doi.org/10.1016/j.cell.2014.11.021 
> Li, G., Fullwood, M. J., Xu, H., Mulawadi, F. H., Velkov, S., Vega, V., Ariyaratne, P. N., Mohamed, Y. B., Ooi, H.-S., & Tennakoon, C. (2010). ChIA-PET tool for comprehensive chromatin interaction analysis with paired-end tag sequencing. Genome biology, 11(2), 1-13. https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-2-r22 
> Yu, S. C. Y., Deng, J., Qiao, R., Cheng, S. H., Peng, W., Lau, S. L., Choy, L. Y. L., Leung, T. Y., Wong, J., Wong, V. W.-S., Wong, G. L. H., Jiang, P., Chiu, R. W. K., Chan, K. C. A., & Lo, Y. M. D. (2022). Comparison of Single Molecule, Real-Time Sequencing and Nanopore Sequencing for Analysis of the Size, End-Motif, and Tissue-of-Origin of Long Cell-Free DNA in Plasma. Clinical Chemistry, 69(2), 168-179. https://doi.org/10.1093/clinchem/hvac180 
> Lajoie, B. R., Dekker, J., & Kaplan, N. (2015). The Hitchhiker’s guide to Hi-C analysis: Practical guidelines. Methods, 72, 65-75. https://doi.org/https://doi.org/10.1016/j.ymeth.2014.10.031 
> Dotson, G. A., Chen, C., Lindsly, S., Cicalo, A., Dilworth, S., Ryan, C., Jeyarajan, S., Meixner, W., Stansbury, C., & Pickard, J. (2022). Deciphering multi-way interactions in the human genome. Nature communications, 13(1), 1-15. 



## METHOD
In order to filter the artifacts caused by random links and other factors, and to improve the quality of higher-order interaction results, we introduce the hypergraph filtering method.The implementation of hypergraph filtering in PPL is divided into two steps: the first is to decompose all the multiple interactions (hyper edges) into VPCs (normal edges), and make the VPCs matrix. Then, sort all the non-zero values in the matrix and obtain the 85% quantile as the filtering threshold. The 85% quantile of the matrix is used as the threshold value. VPCs below this threshold are considered unreliable, and their weights are changed to 0. Second, each hyperedge (concatemer) is decomposed into a fully connected graph in turn, and then, based on the neighbor matrix constructed in the first step, the edges with the value of 0 in the matrix are deleted to generate an intermediate graph. Finally, all the maximal connected subgraphs in the graph are identified using a depth-first traversal algorithm. All the nodes, i.e. genome segments, in each maximal connectivity subgraph are considered to be neighborable in the same space-time. By this method, the ultra-high-order interactions with the participation artifacts will be split into multiple interactions with higher confidence, which will serve as noise reduction. The use of the 85% quantile as the default threshold in the methodology references the findings of Doston et al. (Dotson et al., 2022). Since the method is designed to run at different resolutions, the tool can also function at higher resolutions, but the running time tends to increase in power with resolution.

<img src="https://github.com/versarchey/PPL-Toolbox/blob/main/Denoise/figs/pipeline.png" width = "30%" align=center />

## USAGE
    
    # Simple usage
    java -cp PPL.jar utils.FilterHyper
        #required
            <Input file: multi-contacts>
            <Output file: multi-contacts filtered>
            <Chrom size file> <binSize, e.g. 1000000>
        #optional
            <Percentile cutoff, default:0.85> 
            <Range, e.g. chr1: 1000000-2000000>

    # Test
    # Check inter-intra chromosome ratio 
    java -cp pore-c_tool.jar utils.StatInterIntra
        <Input file1: multi-contacts > 
        <Output file: stats results>

    # Check inter-intra TAD ratio 
    java -cp pore-c_tool.jar utils.StatByTAD 
        <Input file1: multi-contacts > 
        <Input file2: domainfile (.bed 3 colmns format)> 
        <Output file: stats results>




## TEST 

### Check inter-intra TAD ratio change

<img src="https://github.com/versarchey/PPL-Toolbox/blob/main/Denoise/figs/overlap_TAD.png" width = "50%" align=center />

### Target region check

<img src="https://github.com/versarchey/PPL-Toolbox/blob/main/Denoise/figs/filtered_region.png" width = "80%" align=center />

### Correlation with high-quality Hi-C

<img src="https://github.com/versarchey/PPL-Toolbox/blob/main/Denoise/figs/correlation.png" width = "80%" align=center />

